{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7b351ef-1369-4059-9401-8152a1cedc08",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "232fed08-160d-4362-af1a-8a4010eb8a01",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "unindent does not match any outer indentation level (<tokenize>, line 208)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<tokenize>\"\u001b[0;36m, line \u001b[0;32m208\u001b[0m\n\u001b[0;31m    process_files(inputpath, outputpath)\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m unindent does not match any outer indentation level\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "from sequence_dictionaries import protein_weights\n",
    "from sequence_dictionaries import dna_weights, rna_weights\n",
    "from sequence_dictionaries import protein_letters, dna_letters, rna_letters\n",
    "from sequence_dictionaries import rna_table, dna_table\n",
    "from sequence_dictionaries import rna_table_back\n",
    "from sequence_dictionaries import rna_stop_codons, rna_start_codons, dna_stop_codons, dna_start_codons, dna_complement, rna_complement\n",
    "\n",
    "    \n",
    "class IncorrectSequenceLetter(Exception):\n",
    "    def __init__(self, letter, class_name):\n",
    "        self.letter = letter\n",
    "        self.class_name = class_name\n",
    "    \n",
    "    def __str__(self):\n",
    "        return \"The sequence item %s is not found in the alphabet of class %s\" % (self.letter, self.class_name)\n",
    "\n",
    "class Sequence():\n",
    "    \n",
    "    alphabet = set() #creating an alphabet \n",
    "    \n",
    "    def __init__(self, identifier, sequence, alphabet):\n",
    "        self.__identifier = identifier\n",
    "        self.__sequence = sequence\n",
    "\n",
    "        if not all(letter in alphabet for letter in sequence):\n",
    "            invalid_letter = next(letter for letter in sequence if letter not in alphabet)\n",
    "            raise IncorrectSequenceLetter(invalid_letter, self.__class__.__name__)\n",
    "                \n",
    "        \n",
    "    def get_identifier(self):\n",
    "        return self.__identifier\n",
    "    \n",
    "    def get_sequence(self):\n",
    "        return self.__sequence\n",
    "    \n",
    "    def get_mw(self): \n",
    "        mw = 0\n",
    "        for aa in self.__sequence:\n",
    "            mw += protein_weights[aa]\n",
    "        return mw\n",
    "        \n",
    "        if isinstance(self, ProteinSequence):\n",
    "            weigths_dict = protein_weights\n",
    "        elif isinstance(self, DNASequence):\n",
    "            weigths_dict = dna_weights\n",
    "        elif isinstance(self, RNASequence):\n",
    "            weigths_dict = rna_weights\n",
    "            \n",
    "        return sum(weigths_dict[monomer] for monomer in self.get_sequence())\n",
    "    \n",
    "    def has_subsequence(self, sequence_obj):\n",
    "        return sequence_obj.get_sequence() in self.__sequence\n",
    "    \n",
    "    #defining behaviour for the classes defined\n",
    "    \n",
    "    def __len__(self):\n",
    "        return len(self.get_sequence())\n",
    "    \n",
    "    def __eq__(self, other):\n",
    "        return self.get_sequence() == other.get_sequence()\n",
    "    \n",
    "    def __ne__(self, other):\n",
    "        return self.get_sequence() != other.get_sequence()\n",
    "    \n",
    "    def __add__ (self, other):\n",
    "        if type(self) != type(other): #if the sequences are different type, error message is printed\n",
    "            raise ValueError(f\"The sequences must be the same type\")\n",
    "        \n",
    "        idents = self.get_identifier() + \"+\" + other.get_identifier()\n",
    "        seqs = self.get_sequence() + other.get_sequence()\n",
    "\n",
    "        return self.__class__(idents, seqs, alphabet=self.alphabet)\n",
    "    \n",
    "    #Sequence[i], returning the sequence element at position i. Position 0 is the first one. \n",
    "    def __getitem__(self,res):\n",
    "        return self.get_sequence()[res]\n",
    "    \n",
    "    #it is returned a boolean if the sequence has a specific item\n",
    "    def __contains__(self, item):\n",
    "        return item in self.get_sequence()\n",
    "    \n",
    "    def __hash__(self):\n",
    "        return hash((self.get_identifier(), self.get_sequence()))\n",
    "    \n",
    "    #sorting sequences according to their molecular weight \n",
    "    def __lt__(self, other):\n",
    "        return self.get_mw() < other.get_mw() \n",
    "    def __le__(self, other):\n",
    "        return self.get_mw() <= other.get_mw() \n",
    "    def __gt__(self, other):\n",
    "        return self.get_mw() > other.get_mw() \n",
    "    def __ge__(self, other):\n",
    "        return self.get_mw() >= other.get_mw() \n",
    "    \n",
    "    def __hash__(self):\n",
    "        return hash((self.get_identifier(), self.get_sequence()))\n",
    "\n",
    "class ProteinSequence(Sequence):\n",
    "    alphabet = set(protein_letters)\n",
    "    \n",
    "class NucleotideSequence(Sequence):\n",
    "    \n",
    "    def translate(self):\n",
    "\n",
    "        if isinstance(self, DNASequence):\n",
    "            codon_dict = dna_table\n",
    "            start_codon = dna_start_codons\n",
    "            stop_codon = dna_stop_codons\n",
    "        elif isinstance(self, RNASequence):\n",
    "            codon_dict = rna_table\n",
    "            start_codon = rna_start_codons\n",
    "            stop_codon = rna_stop_codons\n",
    "        else:\n",
    "            raise ValueError('Invalid sequence type')\n",
    "    \n",
    "        sequence = self.get_sequence()\n",
    "        aa_list = []\n",
    "\n",
    "        has_start_index = [sequence.find(codon) for codon in start_codon \\\n",
    "            if codon in sequence]\n",
    "        \n",
    "        if has_start_index:\n",
    "            start_i = min([sequence.find(codon) for codon in start_codon \\\n",
    "                if codon in sequence])\n",
    "            for i in range(start_i, len(sequence), 3):\n",
    "                codon = sequence[i: i+3]\n",
    "                if codon in stop_codon or len(codon) != 3 :\n",
    "                    break\n",
    "                aa_list.append(codon_dict[codon])\n",
    "            \n",
    "            return ''.join(aa_list)\n",
    "\n",
    "        else:\n",
    "            raise ValueError('Object has no start codon')\n",
    "\n",
    "#defining DNASequence class\n",
    "class DNASequence(NucleotideSequence):\n",
    "    alphabet = set(dna_letters)\n",
    "    \n",
    "    def transcribe(self):\n",
    "        return ''.join([dna_complement[DNA] for DNA in self.get_sequence()])\n",
    "\n",
    "#defining RNASequence class\n",
    "\n",
    "class RNASequence(NucleotideSequence):\n",
    "    alphabet = set(rna_letters)\n",
    "\n",
    "    def reverse_transcribe(self):\n",
    "        return ''.join([rna_complement[RNA] for RNA in self.get_sequence()])\n",
    "    \n",
    "    \n",
    "def FASTA_iterator(filename, class_name):\n",
    "    \n",
    "    class_map = {\n",
    "        'ProteinSequence': ProteinSequence,\n",
    "        'DNASequence': DNASequence,\n",
    "        'RNASequence': RNASequence\n",
    "    }\n",
    "    \n",
    "    if class_name not in class_map:\n",
    "        raise ValueError(f\"Invalid class_name: {class_name}. It must be one of {', '.join(class_map.keys())}\")\n",
    "    \n",
    "    sequence_class = class_map[class_name]\n",
    "    \n",
    "    with open(filename, 'r') as fd:\n",
    "        file = fd.read()\n",
    "        chains = file.split(\">\")[1:]\n",
    "    \n",
    "        for chain in chains:\n",
    "            try: \n",
    "            \n",
    "                identifier = chain.split(\"\\n\", 1)[0]\n",
    "                sequence = chain.split(\"\\n\", 1)[1].replace(\"\\n\", \"\")\n",
    "            \n",
    "            \n",
    "                seq_instance = sequence_class(identifier, sequence, alphabet=sequence_class.alphabet)\n",
    "\n",
    "                yield seq_instance\n",
    "            \n",
    "            except IncorrectSequenceLetter as e:\n",
    "                sys.stderr.write(\"ValueError: {}\\n\".format(e))\n",
    "                continue\n",
    "                \n",
    "def process_files(inputpath, outputpath=None):\n",
    "    \n",
    "def main():\n",
    "    \n",
    "    inputpath = None\n",
    "    outputpath = None\n",
    "    \n",
    "    if len(sys.argv) == 1:\n",
    "        inputpath = '.'\n",
    "    elif len(sys.argv) == 2:\n",
    "        inputpath = sys.argv[1]\n",
    "    elif len(sys.argv) == 3:\n",
    "        inputpath = sys.argv[1]\n",
    "        outputpath = sys.argv[2]\n",
    "    else:\n",
    "        print(\"Usage: python script.py [inputpath] [outputpath]\")\n",
    "        sys.exit(1)\n",
    "        \n",
    "    if not os.path.exists(inputpath):\n",
    "        print(f\"Error: Input path '{inputpath}' does not exist.\")\n",
    "        sys.exit(1)\n",
    "        \n",
    "     process_files(inputpath, outputpath)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n",
    "    \n",
    "#Example usage:\n",
    "\n",
    "#Creating instances of sequences\n",
    "#protein_seq1 = ProteinSequence(identifier=\"P1\", sequence=\"MVKLHB\", alphabet=set(protein_letters))\n",
    "#protein_seq2 = ProteinSequence(identifier=\"P2\", sequence=\"MVKLH\", alphabet=set(protein_letters))\n",
    "\n",
    "#protein_seq1 == protein_seq2\n",
    "\n",
    "#print(protein_seq1.get_sequence())\n",
    "    #super(Nucleotide, self).__init__(ProteinSequence) #returning the parent class\n",
    "    \n",
    "# Example usage for Protein sequences\n",
    "protein_iterator = FASTA_iterator('example_fasta_file.fa', 'ProteinSequence')\n",
    "for protein_sequence in protein_iterator:\n",
    "    print(protein_sequence.get_identifier())\n",
    "    print(protein_sequence.get_sequence())\n",
    "    print(protein_sequence.get_mw())\n",
    "\n",
    "# Example usage for DNA sequences\n",
    "#dna_iterator = FASTA_iterator('example_fasta_file.fa', 'DNASequence')\n",
    "#for dna_sequence in dna_iterator:\n",
    "#    print(dna_sequence.get_identifier())\n",
    "#    print(dna_sequence.get_sequence())\n",
    "#    print(dna_sequence.transcribe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "274f253a-fb3d-42e5-8463-7181385d5487",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a4446ee-407b-48db-9d6d-476a41e1a1c0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
