{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c470846c-4f5f-4cbd-9498-885df1dc5fd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#1) Define and implement the following classes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ec734cab-eaf5-4eda-9b03-be803a29fc10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from sequence_dictionaries import protein_weights\n",
    "from sequence_dictionaries import dna_weights, rna_weights\n",
    "#print(protein_weights)\n",
    "from sequence_dictionaries import protein_letters, dna_letters, rna_letters\n",
    "from sequence_dictionaries import rna_table, dna_table\n",
    "from sequence_dictionaries import rna_table_back\n",
    "from sequence_dictionaries import rna_stop_codons, rna_start_codons, dna_stop_codons, dna_start_codons, dna_complement, rna_complement\n",
    "\n",
    "\n",
    "class Sequence():\n",
    "    \n",
    "    alphabet = set() #creating an alphabet \n",
    "    \n",
    "    def __init__(self, identifier, sequence, alphabet):\n",
    "        self.__identifier = identifier\n",
    "        self.__sequence = sequence\n",
    "    \n",
    "        if not set(self.__sequence).issubset(self.alphabet):\n",
    "            invalid = set(self.__sequence) - self.alphabet\n",
    "            raise ValueError(f\"Impossible to create instance: {invalid} not possible\")\n",
    "        \n",
    "    def get_identifier(self):\n",
    "        return self.__identifier\n",
    "    \n",
    "    def get_sequence(self):\n",
    "        return self.__sequence\n",
    "    \n",
    "    def get_mw(self): \n",
    "        mw = 0\n",
    "        for aa in self.__sequence:\n",
    "            mw += protein_weights[aa]\n",
    "        return mw\n",
    "        \n",
    "        if isinstance(self, ProteinSequence):\n",
    "            weigths_dict = protein_weights\n",
    "        elif isinstance(self, DNASequence):\n",
    "            weigths_dict = dna_weights\n",
    "        elif isinstance(self, RNASequence):\n",
    "            weigths_dict = rna_weights\n",
    "            \n",
    "        return sum(weigths_dict[monomer] for monomer in self.get_sequence())\n",
    "    \n",
    "    def has_subsequence(self, sequence_obj):\n",
    "        return sequence_obj.get_sequence() in self.__sequence\n",
    "    \n",
    "    #defining behaviour for the classes defined\n",
    "    \n",
    "    def __len__(self):\n",
    "        return len(self.get_sequence())\n",
    "    \n",
    "    def __eq__(self, other):\n",
    "        return self.get_sequence() == other.get_sequence()\n",
    "    \n",
    "    def __ne__(self, other):\n",
    "        return self.get_sequence() != other.get_sequence()\n",
    "    \n",
    "    def __add__ (self, other):\n",
    "        if type(self) != type(other): #if the sequences are different type, error message is printed\n",
    "            raise ValueError(f\"The sequences must be the same type\")\n",
    "        \n",
    "        idents = self.get_identifier() + \"+\" + other.get_identifier()\n",
    "        seqs = self.get_sequence() + other.get_sequence()\n",
    "\n",
    "        return self.__class__(idents, seqs, alphabet=self.alphabet)\n",
    "    \n",
    "    #Sequence[i], returning the sequence element at position i. Position 0 is the first one. \n",
    "    def __getitem__(self,res):\n",
    "        return self.get_sequence()[res]\n",
    "    \n",
    "    #it is returned a boolean if the sequence has a specific item\n",
    "    def __contains__(self, item):\n",
    "        return item in self.get_sequence()\n",
    "    \n",
    "    def __hash__(self):\n",
    "        return hash((self.get_identifier(), self.get_sequence()))\n",
    "    \n",
    "    #sorting sequences according to their molecular weight \n",
    "    def __lt__(self, other):\n",
    "        return self.get_mw() < other.get_mw() \n",
    "    def __le__(self, other):\n",
    "        return self.get_mw() <= other.get_mw() \n",
    "    def __gt__(self, other):\n",
    "        return self.get_mw() > other.get_mw() \n",
    "    def __ge__(self, other):\n",
    "        return self.get_mw() >= other.get_mw() \n",
    "    \n",
    "    def __hash__(self):\n",
    "        return hash((self.get_identifier(), self.get_sequence()))\n",
    "\n",
    "class ProteinSequence(Sequence):\n",
    "    alphabet = set(protein_letters)\n",
    "    \n",
    "class NucleotideSequence(Sequence):\n",
    "    \n",
    "    def translate(self):\n",
    "\n",
    "        if isinstance(self, DNASequence):\n",
    "            codon_dict = dna_table\n",
    "            start_codon = dna_start_codons\n",
    "            stop_codon = dna_stop_codons\n",
    "        elif isinstance(self, RNASequence):\n",
    "            codon_dict = rna_table\n",
    "            start_codon = rna_start_codons\n",
    "            stop_codon = rna_stop_codons\n",
    "        else:\n",
    "            raise ValueError('Invalid sequence type')\n",
    "    \n",
    "        sequence = self.get_sequence()\n",
    "        aa_list = []\n",
    "\n",
    "        has_start_index = [sequence.find(codon) for codon in start_codon \\\n",
    "            if codon in sequence]\n",
    "        \n",
    "        if has_start_index:\n",
    "            start_i = min([sequence.find(codon) for codon in start_codon \\\n",
    "                if codon in sequence])\n",
    "            for i in range(start_i, len(sequence), 3):\n",
    "                codon = sequence[i: i+3]\n",
    "                if codon in stop_codon or len(codon) != 3 :\n",
    "                    break\n",
    "                aa_list.append(codon_dict[codon])\n",
    "            \n",
    "            return ''.join(aa_list)\n",
    "\n",
    "        else:\n",
    "            raise ValueError('Object has no start codon')\n",
    "\n",
    "#defining DNASequence class\n",
    "class DNASequence(NucleotideSequence):\n",
    "    alphabet = set(dna_letters)\n",
    "    \n",
    "    def transcribe(self):\n",
    "        return ''.join([dna_complement[DNA] for DNA in self.get_sequence()])\n",
    "\n",
    "#defining RNASequence class\n",
    "\n",
    "class RNASequence(NucleotideSequence):\n",
    "    alphabet = set(rna_letters)\n",
    "\n",
    "    def reverse_transcribe(self):\n",
    "        return ''.join([rna_complement[RNA] for RNA in self.get_sequence()])\n",
    "\n",
    "#Example usage:\n",
    "\n",
    "#Creating instances of sequences\n",
    "protein_seq1 = ProteinSequence(identifier=\"P1\", sequence=\"MVKLH\", alphabet=set(protein_letters))\n",
    "protein_seq2 = ProteinSequence(identifier=\"P2\", sequence=\"MVKLH\", alphabet=set(protein_letters))\n",
    "\n",
    "protein_seq1 == protein_seq2\n",
    "    #super(Nucleotide, self).__init__(ProteinSequence) #returning the parent class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c733dd78-1a44-494b-ab21-20a734857902",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b22f8cb-8756-4a54-8ea9-d011907a6fc4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1b28d40-2d70-4ad8-81f7-a28fb0c93cb5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5d28fa8-88e9-4d30-9a7c-749b8e82e009",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3094312d-9a14-41b2-a704-e12a17ed3ec8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5da344c2-4c68-42a9-b0ab-35c7f5209a32",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
